File: Move.py
# -*- coding: utf-8 -*-
# Copyright © 2013 Jens Schmer, Michael Engelhard

class Move(object):
    """description of a move"""
    cols = "abcde"
    rows = "123456"

    def __init__(self, fr, to):
        """
        fr and to are tuples in the form "(x, y)"
        """
        self.start = fr
        self.end   = to

    def __str__(self):
        return self.pos_to_str(self.start) + "-" + self.pos_to_str(self.end)

    def pos_to_str(self, pos):
        return Move.cols[pos[0]-1]+str(pos[1])

    def __eq__(self, other):
        return self.start == other.start and self.end == other.end

    def parse_position(msg):
        if len(msg) > 2: raise ValueError("Too much position information")
        if msg[0] not in Move.cols: raise ValueError("Wrong column index")
        if msg[1] not in Move.rows: raise ValueError("Wrong row index")

        x = Move.cols.index(msg[0])+1
        y = int(msg[1])
        return (x, y)

    def from_string(msg):
        msg = msg.strip()
        start, end = msg.split("-")

        return Move(Move.parse_position(start), Move.parse_position(end))

File: src/Move.cpp
#include "MiniCheese/Move.h"

#include "MiniCheese/Exceptions.h"
#include "MiniCheese/StringHelper.h"

using std::string;

//
// free standing functions
std::string posToString(Pos2D pos) {
    char buf[32];
    sprintf_s(buf, "%d", pos.second);

    return Move::cols.at(pos.first-1) + string(buf);
}

Pos2D parsePosition(std::string str) {
    using String::existIn;

    if (str.length() > 2) throw ValueError("Too much position information");
    if (!existIn(str[0], Move::cols)) throw ValueError("Wrong column index");
    if (!existIn(str[1], Move::rows)) throw ValueError("Wrong row index");

    auto x = Move::cols.find(str[0]) + 1;
    auto y = atoi(&str[1]);
    return Pos2D(x, y);
}

Move parseMove(std::string str) {
    auto idx = str.find("-");
    if (str.length() != 5 || idx == std::string::npos)
        throw ValueError(str + " is not a valid move string!");

    return Move(parsePosition(str.substr(0, 2)), parsePosition(str.substr(3, 2)));
}

//
// class Move
const string Move::cols = "abcde",
             Move::rows = "123456";

Move::Move(Pos2D start, Pos2D end)
    : start(start), end(end)
{
}

Move::~Move()
{
}

std::string Move::toString() const {
    return posToString(start) + "-" + posToString(end);
}

bool Move::operator==(const Move &rhs) {
    return start == rhs.start && end == rhs.end;
}